plugin/src/main/java/org/jenkinsci/plugins/structs/describable/ParameterType.java

```diff
@@ -61,47 +61,47 @@ static ParameterType of(Type type) {
+                for (Class<?> subtype : subtypes) {
+                    String simpleName = subtype.getSimpleName();
+                    List<Class<?>> bySimpleName = subtypesBySimpleName.get(simpleName);
+                    if (bySimpleName == null) {
+                        subtypesBySimpleName.put(simpleName, bySimpleName = new ArrayList<Class<?>>());
                     }
-                    Map<String,DescribableModel<?>> types = new TreeMap<String,DescribableModel<?>>();
-                    for (Map.Entry<String,List<Class<?>>> entry : subtypesBySimpleName.entrySet()) {
-                        if (entry.getValue().size() == 1) { // normal case: unambiguous via simple name
+                    bySimpleName.add(subtype);
+                }
+                Map<String, DescribableModel<?>> types = new TreeMap<String, DescribableModel<?>>();
+                for (Map.Entry<String, List<Class<?>>> entry : subtypesBySimpleName.entrySet()) {
+                    if (entry.getValue().size() == 1) { // normal case: unambiguous via simple name
+                        try {
+                            types.put(entry.getKey(), DescribableModel.of(entry.getValue().get(0)));
+                        } catch (Exception x) {
+                            LOGGER.log(Level.FINE, "skipping subtype", x);
+                        }
+                    } else { // have to diambiguate via FQN
+                        for (Class<?> subtype : entry.getValue()) {
                             try {
-                                types.put(entry.getKey(), DescribableModel.of(entry.getValue().get(0)));
+                                types.put(subtype.getName(), DescribableModel.of(subtype));
                             } catch (Exception x) {
                                 LOGGER.log(Level.FINE, "skipping subtype", x);
                             }
-                        } else { // have to diambiguate via FQN
-                            for (Class<?> subtype : entry.getValue()) {
-                                try {
-                                    types.put(subtype.getName(), DescribableModel.of(subtype));
-                                } catch (Exception x) {
-                                    LOGGER.log(Level.FINE, "skipping subtype", x);
-                                }
-                            }
                         }
                     }
-                    return new HeterogeneousObjectType(c, types);
                 }
+                return new HeterogeneousObjectType(c, types);
             }
-            if (Types.isSubClassOf(type, Collection.class)) {
-                return new ArrayType(type, of(Types.getTypeArgument(Types.getBaseClass(type,Collection.class), 0, Object.class)));
-            }
-            throw new UnsupportedOperationException("do not know how to categorize attributes of type " + type);
```

Comment:
This seems to have been removed and I am not sure why. Where is the fallback?

Reply:
I guess it is just not necessary any more and whatever sanity-checking was needed is already done in `DescribableModel`?

Reply:
The fallback became unreachable, so I removed it. I was surprised that no unit tests failed. There is not much sanity-checking in `DescribableModel` or `DescribableParameter`.

I thought about limiting support for `Homo`/`HeterogeneousObjectType` to `Describable` types and keeping the fallback, but that triggered a test failure in `DescribableModelTest#recursion` since `Recursion` is not a `Describable`. I'm not sure if that is on purpose.